### ISSUE TITLE
[Spec] Phase 4a MLP Surrogate Architecture — Input/Output Design & Training Strategy

### ISSUE THREAD

**Issue by anchapin** [anchapin/fluxion#764]
## Overview

This issue defines the complete MLP surrogate architecture for Phase 4a: feature set, network topology, training strategy, and ONNX export contract. This spec is the canonical reference for both `scripts/train_surrogate.py` (Phase 4a deliverable in #718) and the `SurrogateInputs` struct extension in `src/ai/surrogate.rs`.

> **Architecture decision**: MLP-first (this spec), MLP+GP ensemble as Phase 4b follow-on (unlocks #708).

---

## 1. Current Surrogate State

The existing `SurrogateInputs` struct (6 fields) and `analytical_loads` fallback (just sinusoidal solar gain as a placeholder) are intentionally minimal. The `SurrogateManager` ONNX-rt infrastructure is fully built and ready — this spec fills in what gets loaded into it.

```rust
// Current (minimal)
pub struct SurrogateInputs {
    pub exterior_temp: f64,   // -50..60°C
    pub zone_temp: f64,       // 10..40°C
    pub solar_rad: f64,       // 0..1200 W/m²
    pub humidity: f64,        // 0..100%
    pub occupancy: f64,       // 0..10 ppl/100m²
    pub climate_zone: String, // categorical: "4A", "5A", "6A"
}
```

---

## 2. Extended Input Feature Set (Phase 4a)

**Design rationale**: Fluxion uses CTF (Conduction Transfer Function) and FD solvers. The surrogate must capture the same physical drivers: outdoor boundary conditions, interior setpoints, building envelope properties, and temporal thermal lag. Features are grouped by physical origin.

### 2a. Boundary Conditions (4 features)

| Feature | Range | Units | Notes |
|---------|-------|-------|-------|
| `exterior_temp` | -50 to 60 | °C | Already in struct |
| `solar_rad_global` | 0 to 1200 | W/m² | Rename of `solar_rad` |
| `humidity` | 0 to 100 | % RH | Already in struct |
| `wind_speed` | 0 to 20 | m/s | **New** — affects infiltration + convective film coefficient |

### 2b. Interior Conditions (3 features)

| Feature | Range | Units | Notes |
|---------|-------|-------|-------|
| `zone_temp_setpoint` | 10 to 40 | °C | Rename of `zone_temp` |
| `occupancy_density` | 0 to 10 | ppl/100m² | Already in struct |
| `internal_gains` | 0 to 50 | W/m² | **New** — lighting + equipment |

### 2c. Temporal Features (4 features — cyclical encoding)

Critical for capturing thermal lag and diurnal/seasonal variation in CTF outputs.

| Feature | Encoding | Notes |
|---------|----------|-------|
| `hour_sin` | sin(2π · h / 24) | **New** |
| `hour_cos` | cos(2π · h / 24) | **New** |
| `doy_sin` | sin(2π · doy / 365) | **New** — day-of-year |
| `doy_cos` | cos(2π · doy / 365) | **New** |

### 2d. Building Envelope Properties (6 features)

These are held constant for a given building config and varied during LHC sampling (Phase 3, #719).

| Feature | Range | Units | Notes |
|---------|-------|-------|-------|
| `wall_u_value` | 0.1 to 3.0 | W/m²K | Overall opaque envelope |
| `window_to_wall_ratio` | 0.05 to 0.80 | fraction | |
| `window_shgc` | 0.20 to 0.87 | fraction | Solar heat gain coefficient |
| `floor_area` | 20 to 5000 | m² | log-normalized before training |
| `thermal_mass_index` | 0.1 to 1.0 | fraction | Normalized from CTF Cm coefficient |
| `infiltration_ach` | 0.1 to 3.0 | ACH | At 50 Pa (or natural ACH) |

### 2e. Climate Zone (one-hot, 6 features)

Encode all ASHRAE 169-2021 zones represented in `data/weather_locations.json`. For Phase 4a, expand from the current 3 zones (`4A`, `5A`, `6A`) to the full set present in weather data.

**Total input dimensionality: 23 features** (17 continuous + 6 one-hot for climate zone)

---

## 3. Output Feature Set

| Output | Range | Units | Notes |
|--------|-------|-------|-------|
| `heating_load` | 0 to 500 | kW | Positive when heating demand > 0 |
| `cooling_load` | 0 to 500 | kW | Positive when cooling demand > 0 |

> **Note**: Sign convention — both outputs are non-negative magnitudes. The simulation mode (heating vs. cooling dominant) is implicit in the inputs. Per ASHRAE 140 convention, heating and cooling loads are reported separately.

---

## 4. Network Topology

```
Input (23) → Linear(23→64) → BatchNorm → ReLU
          → Linear(64→64) → BatchNorm → ReLU
          → Linear(64→32) → BatchNorm → ReLU
          → Linear(32→2)  [heating_load, cooling_load]
```

**Design choices**:
- **3 hidden layers, 64/64/32 neurons** — based on literature (ORNL large-scale EnergyPlus surrogate study used similar depth; IBPSA 2024 protocol recommends depth-first search from 2-4 layers)
- **BatchNorm** — improves training stability for building physics inputs with heterogeneous scales
- **ReLU** — standard for regression surrogates; no dropout for deterministic inference
- **No output activation** — raw regression (outputs can be negative during training; clamp to 0 post-inference)
- Total parameters: ~9,000 (fast inference, fits in L1 cache)

**Alternatives to benchmark against Phase 4a baseline**:
- Shallow: 2 layers 128/64 (fewer non-linearities, faster export)
- Deep: 4 layers 128/128/64/32 (more capacity for nonlinear interactions)

---

### Data
- Source: LHC sweep output from Phase 3 (#719) — physics engine runs per sample
- Volume: **50K–200K samples** per `(component × building_type)` combination (per IBPSA protocol: ANN needs ≥10K; GP performs well with 1K–8K → MLP targets 50K for robustness)
- Split: 70% train / 15% validation / 15% test (reproducible, seed=42)

### Normalization
- **Inputs**: `StandardScaler` (zero mean, unit variance) per feature — fit on train split only
- **Outputs**: `StandardScaler` on `[heating_load, cooling_load]` — required because cooling >> heating in summer and vice versa
- **Important**: Scaler parameters (`mean_`, `scale_`) must be embedded as constants in the ONNX graph so the Rust `SurrogateManager` receives raw physical units

### Optimizer & Schedule
- Optimizer: Adam (lr=1e-3, betas=(0.9, 0.999))
- LR schedule: CosineAnnealingLR (T_max=100 epochs)
- Early stopping: patience=50 epochs on validation MSE
- Batch size: 512

### Loss
- Primary: MSE on normalized outputs
- Monitoring: RMSE% (normalized RMSE as % of output range) — **acceptance threshold: < 2%**

---

## 6. ONNX Export Contract

```python
torch.onnx.export(
    model,
    dummy_input,                    # shape: (1, 23)
    "models/surrogate_zone.onnx",
    opset_version=17,               # ort 2.0.0-rc.10 compatible
    input_names=["inputs"],
    output_names=["loads"],
    dynamic_axes={"inputs": {0: "batch"}, "loads": {0: "batch"}},
    # Scaler constants embedded as graph initializers
)
```

**Rust-side contract** (what `SurrogateManager::load_onnx` sees):
- Input tensor: `float32[batch, 23]` — raw physical units (scaler embedded in graph)
- Output tensor: `float32[batch, 2]` — `[heating_load_kW, cooling_load_kW]`
- Both inputs and outputs in physical units — no Rust-side normalization needed

---

## 7. `SurrogateInputs` Struct Extension

The Rust struct needs to be extended to match the 23-feature input. Proposed extension (backwards-compatible via `Default`):

```rust
pub struct SurrogateInputs {
    // Boundary conditions
    pub exterior_temp: f64,        // °C
    pub solar_rad_global: f64,     // W/m² (renamed)
    pub humidity: f64,             // %
    pub wind_speed: f64,           // m/s (new)
    // Interior
    pub zone_temp_setpoint: f64,   // °C (renamed)
    pub occupancy_density: f64,    // ppl/100m² (renamed)
    pub internal_gains: f64,       // W/m² (new)
    // Temporal
    pub hour_sin: f64,             // cyclical hour (new)
    pub hour_cos: f64,
    pub doy_sin: f64,              // cyclical day-of-year (new)
    pub doy_cos: f64,
    // Envelope
    pub wall_u_value: f64,         // W/m²K (new)
    pub window_to_wall_ratio: f64, // fraction (new)
    pub window_shgc: f64,          // fraction (new)
    pub floor_area: f64,           // m² (new)
    pub thermal_mass_index: f64,   // fraction (new)
    pub infiltration_ach: f64,     // ACH (new)
    // Climate
    pub climate_zone: String,      // ASHRAE 169 zone string (expanded)
}
```

> **Migration note**: `SurrogateDomain::is_valid` needs updating to validate the new fields. The old `from_temps` constructor should remain for backwards compat with tests, defaulting new fields to mid-range values.

---

## 8. Acceptance Criteria

- [ ] Spec reviewed and approved (this issue)
- [ ] `SurrogateInputs` extended per Section 7 (Rust PR)
- [ ] `scripts/generate_training_data.py` produces dataset matching Section 2 feature schema
- [ ] `scripts/train_surrogate.py` trains to < 2% RMSE on test split
- [ ] `scripts/export_onnx.py` exports with opset 17, input/output shapes per Section 6
- [ ] Rust integration test: `SurrogateManager::load_onnx("models/surrogate_zone.onnx")` succeeds and `predict_loads_governed` uses neural path

---

## 9. Related

- Parent epic: #718 (v3.0 Surrogate Training & ONNX Export)
- Data source: #719 (v2.1 Synthetic Data Generation)
- MIRAI downstream: #708 (temporal forecasting — Phase 4b GP ensemble adds UQ needed here)
- Research: IBPSA 2024 protocol for NN surrogate development; ORNL large-scale EnergyPlus surrogate study; PINN literature being tracked by Research & Innovation Engineer

**Comment by anchapin** [comment]
## Research Coordination: Literature Scan Findings Relevant to MLP Architecture Spec

*Research & Innovation Engineer — May 12, 2026*

Full findings posted on #718 (PINN + ONNX pipeline) and #719 (synthetic data + LHC bounds). Below is the distilled set relevant to the Phase 4a MLP architecture spec.

---

## Three Things Confirmed for #764

### 1. Phase 4a MLP architecture is PINN-compatible with no changes needed

The PINN physics-informed loss is **injected at the loss function level, not the architecture level**. The standard MLP spec remains valid. Adding physics-informed training for Phase 4b requires only:
- Adding an ODE residual loss term (the 5R1C equations from `thermal_model_core.rs`)
- Exposing a `λ_physics` hyperparameter
- No changes to layer dimensions, activations, or ONNX export format

### 2. Training set size: 50K is well-supported; no need to go to 200K for Phase 4a

From IBPSA 2024 protocol survey of 68 papers:
- Typical range: 150–50,000 samples
- 50K is conservative and appropriate for a 10–20 parameter input space
- If PINN physics loss is added in Phase 4b: 5K–15K would suffice

**Recommendation**: Target 50K for Phase 4a. Reserve 200K as a fallback only if accuracy analysis shows underfitting after initial training run.

### 3. λ_physics sensitivity analysis (for Phase 4b planning)

From IBPSA 2025 paper: λ_physics is a **critical hyperparameter** that needs systematic tuning:
- Optimal range typically 0.1–0.3 for building thermal applications
- Sensitivity analysis over λ = [0, 0.05, 0.1, 0.2, 0.5, 1.0] is recommended before settling on a value
- Should be included in the Optuna/hyperparameter search space for Phase 4b

---

## CTF Outputs as Training Targets — Confirmed Viable

fluxion's CTF solver outputs (hourly surface heat fluxes, zone temperatures) are well-suited as MLP training targets:
- CTF outputs satisfy the RC energy balance by construction → consistent training signal
- Deterministic, reproducible → no measurement noise
- Limitation: PINN trained on CTF outputs inherits CTF errors for high-mass constructions (900-series). Phase 4b should migrate to FD outputs once #726 is resolved.

---

See #718 and #719 for full PINN literature scan, ONNX/Rust pipeline spec, LHC parameter bounds, and sampling methodology.

**Comment by anchapin** [comment]
## Literature Calibration Update (2026-05-12)

Based on the Research & Innovation Engineer's scan of the IBPSA 2024 systematic protocol paper (68 BEM surrogate studies) and LHC parameter bounds literature, the following corrections apply to this spec.

### Section 2d — Corrected Envelope Bounds

Original bounds in this spec were too wide for realistic residential/commercial building stock. Literature-calibrated values:

| Feature | Original Range | **Calibrated Range** | Key Note |
|---------|---------------|---------------------|----------|
| `wall_u_value` | 0.1–3.0 W/m²K | **0.18–0.50 W/m²K** | Tighter; covers code-min to high-perf |
| `window_shgc` | 0.20–0.87 | **0.22–0.62** | Most influential parameter overall |
| `window_to_wall_ratio` | 0.05–0.80 | **0.10–0.60** | ±5% WWR → ~50% cooling load swing |
| `infiltration_ach` | 0.1–3.0 ACH | **0.3–2.0 ACH** | Below 0.3 is passive-house territory |

`floor_area`, `thermal_mass_index`, and boundary condition bounds remain unchanged.

### Section 5 — Training Volume

IBPSA 2024 protocol (68 studies, typical range 150–50K samples): **50K is well-supported for 17–23 input params**. The upper bound of 200K in the original spec can be dropped — only needed if underfitting is observed. Revised target: **50K samples**.

### New: ASHRAE 140 Data Isolation Requirement

> **Critical**: The exact parameter configurations used in ASHRAE 140 BESTEST (Standard 140 Annex B cases) must be placed in a **held-out validation split, not training data**. This prevents the surrogate from memorizing the validation cases and gives a clean measure of generalization to the standard test suite.

This means: before running the LHC sweep (#719), serialize the exact ASHRAE 140 case configs as `data/ashrae140_holdout.json`. The LHC sampler must exclude these configurations from the training draw.

### Phase 4b: PINN Compatibility Confirmed

The 5R1C ODE physics loss is a **loss-function add-on only** — no changes to the MLP architecture in Section 4 are required for Phase 4b compatibility. PINN path would reduce training data requirement to 5K–15K samples and improve physical consistency at extrapolation boundaries. Tracking in #718.

### Updated Section 5 (Training Strategy)

Revised data volume table:

| Phase | Approach | Training Samples | Notes |
|-------|----------|-----------------|-------|
| 4a | Pure MLP | **50K** | This spec |
| 4b | MLP + PINN loss | 5K–15K | Reduced due to physics constraint |
| 4b | MLP + GP ensemble | 50K | Same data, GP head on top |

**Comment by anchapin** [comment]
## SPEC AMENDMENT — Modular Architecture Evaluation (from #1178)

**Source:** Research digest Jun 22 → #1178 (ACM e-Energy 2025: "Physics-Informed Modularized Neural Networks for Building Dynamic Modeling")

---

### Change 1: Add modular architecture as a Phase 4a design candidate

The current spec describes a **monolithic MLP** (23 inputs → single network → 2 outputs). This amendment adds a **modular MLP** variant as a parallel candidate to be benchmarked before committing to the final Phase 4a architecture.

#### Proposed module decomposition

| Module | Inputs | Outputs | Physics constraint |
|--------|--------|---------|-------------------|
| **A — Envelope** | wall_u_value, window_to_wall_ratio, window_shgc, floor_area, thermal_mass_index, orientation, climate_zone, exterior_temp, solar_rad_global, wind_speed, hour_sin/cos, doy_sin/cos | conduction + radiation heat flows | Thermal balance (energy in = energy stored + transmitted) |
| **B — HVAC** | zone_temp_setpoint, heating/cooling setpoints, equipment specs (incl. min_heating_power_kw, min_cooling_power_kw), climate_zone, exterior_temp | heating_load, cooling_load (kW) | COP curves; **mechanical minimum power bounds as hard output constraints** (see also #718 amendment) |
| **C — Internal Gains** | occupancy_density, internal_gains (W/m²), plug load density, lighting density | sensible gains, latent gains | Occupancy/schedule mass balance |

Physics constraints are applied **locally within each module** rather than as global loss terms — the key architectural difference from a PINN-augmented monolithic MLP. MLP-only priority is preserved: all modules train MLP-only first; PINN loss augmentation and GP ensemble are Phase 4b.

#### Why this fits fluxion
The thermal network and HVAC system are already separate solvers in fluxion. The modular NN architecture mirrors this directly, making input→module mapping natural and reducing the risk of cross-subsystem spurious correlations during training.

---

### Change 2: Benchmark requirement before architecture commit

Before Phase 4a training begins, run a structured benchmark:

| Metric | Monolithic MLP | Modular MLP |
|--------|---------------|-------------|
| RMSE on heating_load | TBD | TBD |
| RMSE on cooling_load | TBD | TBD |
| Inference time (Rust, ort) | TBD | TBD |
| Retraining time (single subsystem change) | N/A | TBD |

**Decision rule:** If modular ≤ monolithic on RMSE and inference time within 10%, adopt modular as primary architecture.

---

### Change 3: ONNX export contract update (conditional on modular adoption)

If modular wins benchmark:
- Export yields **3 ONNX files**: `surrogate_envelope.onnx`, `surrogate_hvac.onnx`, `surrogate_gains.onnx`
- `burn-onnx 0.21` supports multi-model no_std pipelines — evaluate as export path
- Rust `SurrogateManager` instantiates 3 `ComponentSurrogate` instances (fits the existing `CompositeSurrogate` pattern)
- Per-module inference can be parallelized if zone-level decomposition allows

---

### Change 4: Phase 4b GP ensemble preview

If Phase 4b GP ensemble is pursued (#718), modular base models allow **per-subsystem uncertainty quantification** (envelope uncertainty ≠ HVAC uncertainty) — a strict improvement over a global GP layer on a monolithic MLP.

---

_No implementation action required yet. Architecture decision deferred to benchmark step above._
